pytensor/scan/rewriting.py

```diff
@@ -1103,6 +1111,71 @@ def sanitize(x):
         return at.as_tensor_variable(x)
 
 
+@node_rewriter([Scan])
+def while_scan_merge_subtensor_last_element(fgraph, scan_node):
+    """
+    Replace while_scan_out[abs(min(tap)):][-1] by while_scan_out[-1], for
+    recurring outputs, asserting that at least one step occurs.
+    Only the first step can be ensured by the inputs alone (i.e., `n_steps > 0`),
+    as the while scan could abort earlier anytime after that. This means it is
+    not possible to replace while_scan_out[abs(min(tap)):][-i]
+    by while_scan_out[-i], for -i != -1.
+    """
+    op = scan_node.op
+
+    if not op.info.as_while:
+        return None
+
+    # Optimization is not implemented form mit-mot
+    recurrent_outputs = op.outer_mitsot_outs(scan_node.outputs) + op.outer_sitsot_outs(
+        scan_node.outputs
+    )
+    recurrent_outputs_taps_slices = (
+        op.info.mit_sot_in_slices + op.info.sit_sot_in_slices
+    )
+
+    n_steps = scan_node.inputs[0]
+    non_zero_steps_cond = n_steps > 0
+    assert_non_zero_steps_op = Assert("n_steps > 0")
+
+    subtensor_merge_replacements = {}
+
+    # Iterate over all nodes that are two computations below the while scan
+    for node in get_clients_at_depth(fgraph, scan_node, depth=2):
+        if not isinstance(node.op, Subtensor):
+            continue
+
+        u = node.inputs[0]
+        if not (u.owner and isinstance(u.owner.op, Subtensor)):
```

Comment:
Not for clients. `u.owner` can only be `None` for input variables, but since these are clients of the scan they must not be input variables by definition.

Still I like to always check anyway. That's what the `u.owner and ...` does. Were you suggesting I remove that first check?

Reply:
Ok, that's what I thought, I wondered why you checked and couldn't think of a way that it could be `None` here. If it cannot be `None` here perhaps an `assert` would be more appropriate, to better convey the intent? It is just a tiny detail, so if you don't feel like changing, I don't mind if you don't :)

Reply:
I was wrong, it can be `none`, because a Subtensor can be a client of scan through an input other than the first (as an index, and not what is being indexed), even though that is unlikely.